src/main/java/net/mcreator/preferences/entries/FileEntry.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * MCreator (https://mcreator.net/)
+ * Copyright (C) 2012-2020, Pylo
+ * Copyright (C) 2020-2023, Pylo, opensource contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package net.mcreator.preferences.entries;
+
+import com.google.gson.JsonElement;
+import javafx.stage.FileChooser;
+import net.mcreator.preferences.PreferencesEntry;
+import net.mcreator.preferences.PreferencesManager;
+import net.mcreator.ui.component.JFileSelector;
+
+import javax.swing.*;
+import java.awt.*;
+import java.io.File;
+import java.util.EventObject;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+public class FileEntry extends PreferencesEntry<File> {
+
+	private final transient boolean isFolder;
+	private final transient FileChooser.ExtensionFilter[] filters;
+	private final transient boolean allowNullValue;
+
+	/**
+	 * The constructor used to add a new preference entry to select the location of a folder.
+	 *
+	 * @param id    The preference entry's ID
+	 * @param value The default value of the entry
+	 */
+	public FileEntry(String id, File value, boolean allowNullValue) {
+		super(id, value);
+		this.isFolder = true;
+		this.allowNullValue = allowNullValue;
+		this.filters = new FileChooser.ExtensionFilter[] {};
+	}
+
+	/**
+	 * The constructor used to add a new preference entry to select the location of a file
+	 *
+	 * @param id      The preference entry's ID
+	 * @param value   The default value of the entry
+	 * @param filters One or multiple filters to apply to the {@link FileChooser} dialog
+	 */
+	public FileEntry(String id, File value, boolean allowNullValue, FileChooser.ExtensionFilter... filters) {
+		super(id, value);
+		this.isFolder = false;
+		this.allowNullValue = allowNullValue;
+		this.filters = filters;
+	}
+
+	@Override public JComponent getComponent(Window parent, Consumer<EventObject> fct) {
+		if (isFolder)
+			return new JFileSelector(parent, value, allowNullValue, fct);
+		else
+			return new JFileSelector(parent, value, allowNullValue, fct, filters);
+	}
+
+	@Override public void setValueFromComponent(JComponent component) {
+		File file = ((JFileSelector) component).getFile();
+		// we don't use null as the old value is kept
+		this.value = Objects.requireNonNullElseGet(file, () -> new File("Not specified"));

```

Comment:
Basing on the previous comments, this should look like this, right?
```suggestion
		this.value = file;
```

Reply:
It doesn't fix the problem with the null value.

Reply:
> null is not serialized, but that is not a problem, as if field is not present in json, it will be properly deserialized back to null

Reply:
Ok, so if this is not a problem, can I know why this doesn't work? I'm doing exactly what you ask me to do.,.. I'm only using null and as I explained before, it doesn't work because null is not serialized and the behaviour of PreferencesManager in this case is to use the default value of the preference entry.

The preference file:
![image](https://github.com/MCreator/MCreator/assets/38427877/9d0b8fbf-b224-449d-803e-5eb4af975f6a)

The UI after restarting MCreator (once you saved the preferences with a FileEntry set to null)
![image](https://github.com/MCreator/MCreator/assets/38427877/332d251c-6553-4a05-b429-8780ff46b873)

Reply:
I see two options:

1. Make file entry have default value of null if it is nullable. In other words, if default value is set to null, field supports null value. I think this covers the majority of use cases of null value. This solves the use of default value mentioned above.
2. If you want the nullable field to also support non-null default values, you could make new wrapper object eg. NullableFile that has two fields: file and isNull that specified if the file is null

I would personally go with 1 as it is more simple and cleaner

Reply:
I tried something to allow null values inside PreferencesManager (see https://github.com/MCreator/MCreator/pull/3946/commits/f29881651dcc37cc6b630409e758fb0f89ffb843), so each PreferenceEntry can define a `allowNullValue` variable. When PreferencesManager is loading preferences, it checks if the PreferenceEntry allows null as a value.

Reply:
This will work the same as option 1. already so I don't think you actually need this refactor as allow null value effectively limits allowed default value to "null" as null becomes valid value and thus can't be identified as "missing value, use default"

Reply:
> I think this covers the majority of use cases of null value.

And what do we do with the other cases? For example, I want to add a Preference to select a folder (with a folder as a default value), but I want this field to also be nullable. It won't be possible. I'm sure it will happen someday, maybe in 2 months or in 2 years, but it will happen eventually and we will have the same problem as we currently have.

Reply:
In this case go with option 2, as the commit you made still does not allow nullable values and non-null default values at the same time

Reply:
It can be defined by each entry individually (like it was before the commit). The value is defined inside the PreferenceEntry's constructor, so it's passed inside the super, making it possible to define it individually... except if there's something I don't know or don't understand.

Reply:
Could be. Open new PR for this separately and I will check it when I am back. I will run some tests, maybe the logic will work as it should indeed

Reply:
Sorry for being such an idiot and blind 😅 I finally understood what you meant with your option #1 and I applied it... and this fixes all problems :)

Reply:
Option 1 still causes this problem you highlighted:

> And what do we do with the other cases? For example, I want to add a Preference to select a folder (with a folder as a default value), but I want this field to also be nullable. It won't be possible. I'm sure it will happen someday, maybe in 2 months or in 2 years, but it will happen eventually and we will have the same problem as we currently have.

And the current solution still uses invalid value `new File("Not specified")` as special file marker which is really hacky and the reason we opened this conversation at the first place.

This is why I said open new PR for support of nullable values as you suggested in https://github.com/MCreator/MCreator/commit/f29881651dcc37cc6b630409e758fb0f89ffb843 and we can check separately how this works out with non-null default values on nullable fields and we can work from there if this will work.

Reply:
Oh... sorry. I tried to remember and read everything, but I guess too much time passed.

However, if nullable gives too much trouble, why don't we simply make it once this PR is merged? I never needed a nullable vslue I think (at least this is not needed for the other PR).

Reply:
It is needed for the java_home parameter sort of

I would say open another PR for nullable handling as suggested in https://github.com/MCreator/MCreator/commit/f29881651dcc37cc6b630409e758fb0f89ffb843 (without file entry specifics) and make sure it handles well cases where:

* field is nullable and the default value is null
* field is nullable and the default value is not null

Then we merge that PR and then this PR